fix Vue node widgets should be in disabled state if their slots are connected with a link

christian-byrne opened the pull request:
## Summary

Fixes https://github.com/Comfy-Org/ComfyUI_frontend/issues/5692 by making widget link connection status trigger on change so Vue widgets with connected links could properly switch to the `disabled` state when they are implicitly converted to inputs. 

## Changes

- **What**: Added `node:slot-links:changed` event tracking and reactive slot data synchronization for Vue widgets

```mermaid
graph TD
    A[Widget Link Change] --> B[NodeInputSlot.link setter]
    B --> C{Is Widget Input?}
    C -->|Yes| D[Trigger slot-links:changed]
    C -->|No| E[End]
    D --> F[Graph Event Handler]
    F --> G[syncNodeSlotData]
    G --> H[Update Vue Reactive Data]
    H --> I[Widget Re-render]
    
    style A fill:#f9f9f9,stroke:#333,color:#000
    style I fill:#f9f9f9,stroke:#333,color:#000
```

## Review Focus

Widget reactivity performance with frequent link changes and event handler memory management in graph operations.

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-5834-fix-Vue-node-widgets-should-be-in-disabled-state-if-their-slots-are-connected-with-a-link-27c6d73d365081f6a6c3c1ddc3905c5e) by [Unito](https://www.unito.io)

DrJKL commented on src/lib/litegraph/src/node/NodeInputSlot.ts:
```diff
@@ -41,11 +42,37 @@ export class NodeInputSlot extends NodeSlot implements INodeInputSlot {
+      slotType: NodeSlotType.INPUT,
+      slotIndex: index,
+      connected: this._link != null,
+      linkId: this._link ?? null
```
```suggestion
      linkId: this.link
```

DrJKL commented on src/lib/litegraph/src/node/NodeInputSlot.ts:
```diff
@@ -41,11 +42,37 @@ export class NodeInputSlot extends NodeSlot implements INodeInputSlot {
+    const graph = this.node.graph
+    if (!graph) return
+
+    const index = this.node.inputs.indexOf(this)
+    if (index === -1) return
+
+    graph.trigger('node:slot-links:changed', {
+      nodeId: this.node.id,
+      slotType: NodeSlotType.INPUT,
+      slotIndex: index,
```
option:
```suggestion
    const slotIndex = this.node.inputs.indexOf(this)
    if (slotIndex === -1) return

    graph.trigger('node:slot-links:changed', {
      nodeId: this.node.id,
      slotType: NodeSlotType.INPUT,
      slotIndex,
```

DrJKL commented on src/composables/graph/useGraphNodeManager.ts:
```diff
@@ -66,6 +66,17 @@ export function useGraphNodeManager(graph: LGraph): GraphNodeManager {
+
+    vueNodeData.set(nodeId, {
+      ...currentData,
+      inputs: node.inputs ? [...node.inputs] : undefined,
```
Not for this PR, but what if we defined inputs and outputs as being required, so that a lack of them would just be an empty array? Would that simplify some of the logic?

christian-byrne committed:
trigger on link change

christian-byrne committed:
emit when widget linked

webfiltered reviewed (commented):
I feel like this might be around 40K lines-too-many.  Give or take.  😁

christian-byrne committed:
update test snapshot for onTrigger property

webfiltered reviewed (commented):
Most importantly: LGTM!

I'm still a little confused about why we're using the `LGraph` "trigger" system - the reason I mention that is just because it feels like we have added a few iterations of  complexity and workarounds.

If we need to revisit again, maybe we should go more nuclear on the fix, and remove any possibility of decade-old design decisions haunting us any further?

DrJKL commented on src/renderer/extensions/vueNodes/components/NodeWidgets.vue:
```diff
@@ -129,12 +130,21 @@ const processedWidgets = computed((): ProcessedWidget[] => {
 
     const vueComponent = getComponent(widget.type) || WidgetInputText
 
+    const slotMetadata = widget.slotMetadata
```
Could pull off options at the same time

christian-byrne committed:
implement proper typed event system with discriminated unions

- Use discriminated union pattern for type-safe events without assertions
- Preserve monkey-patching/chaining for multiple system compatibility
- Convert old (action, param) format to new {type, ...} format in trigger()
- All handlers receive properly typed LGraphTriggerEvent objects

christian-byrne commented:
> I'm still a little confused about why we're using the LGraph "trigger" system - the reason I mention that is just because it feels like we have added a few iterations of complexity and workarounds.
> 
> If we need to revisit again, maybe we should go more nuclear on the fix, and remove any possibility of decade-old design decisions haunting us any further?

I think everyone who reads this all wants to go nuclear on the fix, but it doesn't fully align with business goals to dedicate a lot of time to doing it. We will gradually replace the system with signals-based implementation.

christian-byrne commented:
Planning on making changes:

- Properly type events to not require any assertions or awkward reflection, using standard patterns
- Add e2e tests for the disabled state on linked widgets

christian-byrne committed:
remove unused export

christian-byrne committed:
add aria label

christian-byrne committed:
add browser test for disabled linked widgets

christian-byrne committed:
optimize update

christian-byrne committed:
use named NodeId type

christian-byrne committed:
add comment explaining the disabling business logic

christian-byrne committed:
refactor trigger event type check

christian-byrne commented on src/lib/litegraph/src/LGraphNode.ts:
```diff
@@ -3009,6 +3019,15 @@ export class LGraphNode
         // remove there
         input.link = null
+        if (input.widget) {
+          graph.trigger('node:slot-links:changed', {
+            nodeId: target.id,
+            slotType: NodeSlotType.INPUT,
+            slotIndex: link_info.target_slot,
+            connected: false,
+            linkId: link_info.id
+          })
```
I think it would only be simplified if it were to be made a class instance property of some sort, which would start to go backwards a bit. If it's moved to a more pure helper, then it becomes even more code, I think. Don't see a way to simplify/refactor cleanly.

christian-byrne commented on src/composables/graph/useGraphNodeManager.ts:
```diff
@@ -66,6 +79,22 @@ export function useGraphNodeManager(graph: LGraph): GraphNodeManager {
+
+    if (!nodeRef || !currentData) return
+
+    const refreshedData = extractVueNodeData(nodeRef)
```
https://github.com/Comfy-Org/ComfyUI_frontend/pull/5834/commits/07de517caff3019ea1418f6085c9ab78ff9b614d

christian-byrne commented on src/composables/graph/useGraphNodeManager.ts:
```diff
@@ -444,63 +475,64 @@ export function useGraphNodeManager(graph: LGraph): GraphNodeManager {
+      'node:slot-errors:changed': (slotErrorsEvent) => {
+        refreshNodeSlots(String(slotErrorsEvent.nodeId))
+      },
+      'node:slot-links:changed': (slotLinksEvent) => {
```
No, some things should not be debounced.

christian-byrne commented on src/renderer/extensions/vueNodes/components/NodeWidgets.vue:
```diff
@@ -129,12 +130,21 @@ const processedWidgets = computed((): ProcessedWidget[] => {
+    const slotMetadata = widget.slotMetadata
+
+    let widgetOptions = widget.options
+    if (slotMetadata?.linked) {
```
https://github.com/Comfy-Org/ComfyUI_frontend/pull/5834/commits/eff84313dd841127218ae75d80ff18c860729e06

christian-byrne commented on src/lib/litegraph/src/LGraph.ts:
```diff
@@ -1180,8 +1191,33 @@ export class LGraph
+    param: LGraphTriggerParam<A>
+  ): void
+  trigger(action: string, param: unknown): void
   trigger(action: string, param: unknown) {
```
https://github.com/Comfy-Org/ComfyUI_frontend/pull/5834/commits/3e542a0d0b4cfca359b6e5868dc4fb124f55f2dd
